src/mainboard/pcengines/apu2/mainboard.c

```diff
@@ -259,6 +265,129 @@ static void mainboard_final(void *chip_info)
 	//
 	write_gpio(GPIO_58, 1);
 	write_gpio(GPIO_59, 1);
+
+#if CONFIG_BOARD_PCENGINES_APU2 || CONFIG_BOARD_PCENGINES_APU3 || CONFIG_BOARD_PCENGINES_APU4
+	if (!check_console()) {
+
+		//
+		// The console is disabled, check if S1 is pressed and enable if so
+		//
+		if (!read_gpio(GPIO_32)) {
+
+			const struct spi_flash *flash = NULL;
+			struct region_device rdev;
+			const struct region_device *boot_dev;
+			struct cbfs_props props;
+			struct cbfsf fh;
+			size_t fsize, offset;
+			uint32_t ftype = CBFS_TYPE_RAW;
+			const char* bootorder_name = "bootorder";
+			char* bootorder_file = NULL;
+
+			printk(BIOS_INFO, "S1 PRESSED\n");
+
+			/* get SPI flash boot device */
+			flash = boot_device_spi_flash();
+			if (flash == NULL) {
+				printk(BIOS_WARNING, "Can't get boot flash device\n");
+				return;
+			}
+
+			/* initialize cbfs region device */
+
+			boot_dev = boot_device_ro();
+
+			if (boot_dev == NULL) {
+				printk(BIOS_WARNING, "Can't init CBFS boot device\n");
+				return;
+			}
+
+			if (cbfs_boot_region_properties(&props)) {
+				printk(BIOS_WARNING, "Can't locate CBFS\n");
+				return;
+			}
+
+			if (rdev_chain(&rdev, boot_dev, props.offset, props.size)) {
+				printk(BIOS_WARNING, "Rdev chain failed\n");
+				return;
+			}
+
+			/* locate bootorder file and calculate its offset */
+
+			if (cbfs_locate(&fh, &rdev, bootorder_name, &ftype)) {
+				printk(BIOS_WARNING, "Can't locate file in CBFS\n");
+				return;
+			}
+
+			fsize = region_device_sz(&fh.data);
+			offset = rdev_relative_offset(boot_dev, &fh.data);
+
+			if (fsize & 0xFFF) {
+				printk(BIOS_WARNING,"The bootorder file is not 4k aligned\n");
+				return;
+			}
+
+			bootorder_file = (char *) rdev_mmap(&fh.data, 0, fsize);
+
+			if (!bootorder_file){
+				printk(BIOS_WARNING, "Could not mmap bootorder\n");
+				return;
+			}
+
+			/* find and modify scon knob */
+
+			int index = 0;
+			const char* pattern = "scon";
+			char *lpattern = (char *) pattern;
+			char *bootorder_copy = malloc(fsize);
+
+			bootorder_copy = (char *) memcpy(bootorder_copy, bootorder_file, fsize);
+			rdev_munmap(&rdev, bootorder_file);
+
+			while (*bootorder_copy && *pattern ) {
+				if ( *lpattern == 0)  // the pattern matches return the pointer
+					break;
+				if ( *bootorder_copy == 0) { // We're at the end of the file content but don't have a patter match yet
+					bootorder_copy = NULL;
+					break;
+				}
+				if (*bootorder_copy == *lpattern ) {
+					// The string matches, simply advance
+					bootorder_copy++;
+					lpattern++;
+					index++;
+				} else {
+					// The string doesn't match restart the pattern
+					bootorder_copy++;
+					index++;
+					lpattern = (char *) pattern;
+				}
+			}
+
+			if(bootorder_copy == NULL) {
+				printk(BIOS_WARNING, "scon pattern not found\n");
+				return;
+			}
+
+			*(bootorder_copy) = 0x31;
+			bootorder_copy -= index;
```

Comment:
@miczyg1 what will happen with this pointer when `bootorder_copy` is NULL? This pointer modification without any check looks dangerous.